js/src/ankerSnapshot.ts

```diff
@@ -0,0 +1,206 @@
+  }
+}
+
+class WormholeParameters {
+  core_bridge_program_id: PublicKey;
+  token_bridge_program_id: PublicKey;
+
+  constructor(data: any) {
+    try {
+      const parsedData = {
+        ...data,
+        core_bridge_program_id: new PublicKey(
+          data.core_bridge_program_id.toArray('le')
+        ),
+        token_bridge_program_id: new PublicKey(
+          data.token_bridge_program_id.toArray('le')
+        ),
+      };
+
+      Object.assign(this, parsedData);
+    } catch (error) {
+      console.log('error', error);
+    }
+  }
+}
+
+class Metrics {
+  swapped_rewards_st_sol_total: BN;
+  swapped_rewards_ust_total: BN;
+
+  constructor(data: Metrics) {
+    Object.assign(this, data);
+  }
+}
+
+// @ts-ignore
+const schema = new Map([
+  [
+    WormholeParameters,
+    {
+      kind: 'struct',
+      fields: [
+        ['core_bridge_program_id', 'u256'],
+        ['token_bridge_program_id', 'u256'],
+      ],
+    },
+  ],
+  [
+    Metrics,
+    {
+      kind: 'struct',
+      fields: [
+        ['swapped_rewards_st_sol_total', 'u64'],
+        ['swapped_rewards_ust_total', 'u64'],
+      ],
+    },
+  ],
+  [
+    Anker,
+    {
+      kind: 'struct',
+      fields: [
+        ['solido_program_id', 'u256'],
+        ['solido', 'u256'],
+        ['b_sol_mint', 'u256'],
+        ['token_swap_pool', 'u256'],
+        ['terra_rewards_destination', ['u8', 20]],
+        ['wormhole_parameters', WormholeParameters],
+        ['metrics', Metrics],
+        ['self_bump_seed', 'u8'],
+        ['mint_authority_bump_seed', 'u8'],
+        ['reserve_authority_bump_seed', 'u8'],
+        ['st_sol_reserve_account_bump_seed', 'u8'],
+        ['ust_reserve_account_bump_seed', 'u8'],
+      ],
+    },
+  ],
+]);
+
+export const getAnker = (ankerInstanceDataBuffer: Buffer) => {
+  const deserialized = deserializeUnchecked(
+    schema,
+    Anker,
+    ankerInstanceDataBuffer
+  );
+
+  return deserialized;
+};
+
+export const getSnapshot = async (
+  connection: Connection,
+  programAddresses: ProgramAddresses
+): Promise<AnkerSnapshot> => {
+  // Record of all the account data required for snapshot
+  const accountsInfoRecord: Record<
+    string,
+    AccountInfo<Buffer | ParsedAccountData> | null
+  > = {
+    [programAddresses.ankerInstanceId.toString()]: null,
+    [programAddresses.solidoInstanceId.toString()]: null,
+  };
+
+  const [stSolReserveAccountAddress] = await PublicKey.findProgramAddress(
+    [
+      programAddresses.ankerInstanceId.toBuffer(),
+      Buffer.from('st_sol_reserve_account'),
+    ],
+    programAddresses.ankerProgramId
+  );
+
+  accountsInfoRecord[stSolReserveAccountAddress.toString()] = null;
+
+  // Iterates over all the required accounts data to be fetched and updates the accountsInfo record
+  // Once, all we have the record of all the accounts data, we can create the snapshot
+  // This way, we avoid torn reads, and fetch all the required data in one go with `getMultipleAccountInfo`
+  while (true) {
```

Comment:
I don’t see a `continue` or `break` in this loop, it looks like it will either throw or return in the first iteration.

If you want to mimick what we do in https://github.com/ChorusOne/solido/blob/3faa01864a444197324d209b4c2810f203781022/cli/common/src/snapshot.rs, I would do something like:

* In the functions that read accounts, instead of getting them directly from the `connection`, read them from a hash map. If the account is not part of the hashmap, throw `AccountsMissingException(List[Pubkey])`.
* In the loop, catch this, add the missing accounts to the ones to fetch in the `GetMultipleAccounts` call, and then `continue`.

Reply:
Yeah this is because in this case, we have all the accounts' addresses already in the `accountsInfo` record/hashmap before the loop starts.
So it fetches all the data in the first iteration and returns the snapshot so no checking and `continue`.

Reply:
But then why is the loop there?

Reply:
Yeah that is there just to follow the process we followed for the Solido Snapshot. We can remove it or probably leave it there for future(any other property needs to be added to the snapshot that requires multi-step reads)

Reply:
It does nothing right now, let’s remove it, we can add it again if it’s needed.